kratos/sources/profiler.cpp

```diff
@@ -0,0 +1,236 @@
+#include <atomic>
+
+
+namespace Kratos::Internals {
+
+
+template <class T>
+Profiler<T>::Item::Item(CodeLocation&& r_location)
+    : mRecursionLevel(0),
+      mCallCount(0),
+      mTime(0),
+      mLocation(std::move(r_location))
+{
+}
+
+
+template <class T>
+typename Profiler<T>::Item& Profiler<T>::Item::operator+=(const Item& r_rhs)
+{
+    mCallCount += r_rhs.mCallCount;
+    mTime += r_rhs.mTime;
+    return *this;
+}
+
+
+template <class T>
+Profiler<T>::Profiler()
+    : Profiler("kratos_profiler_output.json")
+{
+}
+
+
+template <class T>
+Profiler<T>::Profiler(std::filesystem::path&& r_outputPath)
+    : mItemContainerMap(),
+      mpScope(),
+      mOutputPath(std::move(r_outputPath))
+{
+    // "Reserve" thread map to avoid bucket moving later on.
+    const auto number_of_threads = std::thread::hardware_concurrency();
+    std::vector<std::thread> threads;
+    threads.reserve(number_of_threads);
+    std::atomic<std::size_t> thread_counter = 0;
+    for (std::size_t i_thread=0; i_thread<number_of_threads; ++i_thread) {
+        threads.emplace_back([i_thread, &thread_counter, this](){
+            while (thread_counter < i_thread) {} // <== wait until the previous thread finishes
+            mItemContainerMap.emplace(std::this_thread::get_id(), std::list<Item> {});
+            ++thread_counter;
+        });
+    }
+    for (auto& r_thread : threads)
+        r_thread.join();
+    threads.clear();
+
+    // Insert first item measuring the total runtime
+    auto& r_item = this->Create(CodeLocation());
+    mpScope.reset(new Scope(r_item)); // <== no, this doesn't immediately become a dangling reference
+}
+
+
+template <class T>
+typename Profiler<T>::Item& Profiler<T>::Create(CodeLocation&& r_item)
+{
+    auto& r_list = mItemContainerMap[std::this_thread::get_id()];
+    r_list.emplace_back(std::move(r_item));
+    return r_list.back();
+}
+
+
+struct SourceLocationHash
+{
+    std::size_t operator()(const CodeLocation& r_argument) const
+    {
+        std::string string(r_argument.GetFileName());
+        string.append(std::to_string(r_argument.GetLineNumber()));
+        return std::hash<std::string>()(string);
+    }
+};
+
+
+struct SourceLocationEquals
+{
+    bool operator()(const CodeLocation& r_lhs,
+                    const CodeLocation& r_rhs) const
+    {
+        return (std::string(r_lhs.GetFileName()) == std::string(r_rhs.GetFileName())) && (r_lhs.GetLineNumber() == r_rhs.GetLineNumber());
+    }
+};
+
+
+namespace {
+template <class TTimeUnit>
+std::string GetTimeUnit()
+{
+    KRATOS_ERROR << "Unknown time unit";
+}
+
+template <>
+std::string GetTimeUnit<std::chrono::milliseconds>()
+{
+    return "ms";
+}
+
+template <>
+std::string GetTimeUnit<std::chrono::microseconds>()
+{
+    return "us";
+}
+
+template <>
+
+std::string GetTimeUnit<std::chrono::nanoseconds>()
+{
+    return "ns";
+}
+} // unnamed namespace
+
+
+template <class T>
+Profiler<T>::~Profiler()
+{
+    // Stop the total timer
```

Comment:
I would put this function in a ToJson() function so we can call it on demand. (for example in each stage or time step)

Reply:
Yep, that's a good idea.

Also, writing to a file is not MPI safe in general. I'm thinking about appending the file name using the default communicator, but I see that using that is discouraged. Any tips or better ideas on how to address this?

Reply:
I think in this particular case it might be of to use the default data comm

Or can you pass one at construction? That would be the preferred way